docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,243 @@
+    - Full UCA implementation for some Unicode collations, for example: `utf8mb4_900_ai_ci`.
+
+## Proposal
+
+### Collate Interface
+
+There are two basic functions needed for collation `sortKey` calculation:
+  * Function that is used to compare two strings according to the given collation.
+  * Function that is used to make a memory-comparable `sortKey` corresponding to the given string.
+The interface can be defined as following:
+
+```
+type CollAttrFlag int
+
+const (
+    FlagIgnoreCase uint64=1 << iota
+    FlagIgnoreAccent
+    FlagPadding
+)
+
+type CharsetCollation interface {
+    // Compare returns an integer comparing the two byte slices. The result will be 0 if a==b, -1 if a < b, and +1 if a > b.
+    Compare(a, b []byte) (int, error)
+    // CompareString returns an integer comparing the two strings. The result will be 0 if a==b, -1 if a < b, and +1 if a > b.
+    CompareString(a, b string) (int, error)
+    // Key returns the collation key for str, the returned slice will point to an allocation in Buffer.
+    Key(buf *Buffer, str []byte) error
+    // KeyFromString returns the collation key for str, the returned slice will point to an allocation in Buffer.
+    KeyFromString(buf *Buffer, str string) error
+
+    /******* Possible Optimization ********/
+    // FromKey returns the original byte slice from collation key.
+    FromKey(buf *Buffer, str []byte) error
+    // StringFromKey returns the original string from collation key.
+    StringFromKey(str []byte) string
+}
+```
+The interface is quite similar to the Go [collate package](https://godoc.org/golang.org/x/text/collate), it is possible to implement some of the MySQL collation by simply mapping it to an existing Go collation.
+
+### Row Format
+
+The encoding layout of TiDB has been described in our [previous article](https://pingcap.com/blog/2017-07-11-tidbinternal2/#map). The row format should be changed to make it memory comparable, this is important to the index lookup. Basic principle is that all keys encoded should use the `sortKeys` result from `Key()`/`KeyFromString()` function. However, most of the `sortKeys` calculations are not reversible.
+
+#### Option 1
+
+  * For table data, encodings stay unchanged. All strings are compared after decoding with the `Compare()` function.
+  * For table indices, we replace current `ColumnValue` with `sortKey`:
+    - For unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKey
+      Value: rowID
+	```
+    - For non-unique indices:
+    ```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKeys_rowID
+      Value: null
+	```
+
+Pros:
+    The current encoding layout keeps unchanged.
+
+Cons:
+     Since index value can not be recovered from `sortKey`, extra table lookup is needed for current Index-only queries, this may lead to performance regressions. The extra table lookup also requires corresponding changes to the TiDB optimizer.
```

Comment:
It is important to state not just performance regressions within TIDB, but migration issues. MySQL has the capability to do covering indexes on columns with collations.

Reply:
> MySQL has the capability to do covering indexes on columns with collations.

I didn't get your point here...The 'covering index' is about optimization of the SQL compiler, it doesn't change anything to the semantic, so that is all about the performance, right?

Reply:
Got your point, thank you.

Reply:
Adding a covering index is a common SQL optimization method for DBAs. If they add a covering index, and still need to look up table. This is counter-intuitive.